### PR TITLE
[networks] Improve memory allocation pattern 

### PR DESCRIPTION
**Pull request by p-lambert**
### What does this PR do?

This PR refactors some parts of the network-tracer to reduce the number and the bursty pattern of allocations.
In our load test server I witnessed a ~30% drop in bytes allocated/s and a ~20% CPU drop

**bytes allocated/s**
Screen Shot 2021-09-21 at 1 14 47 PM

**CPU cores**
Screen Shot 2021-09-21 at 1 14 25 PM

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

I'm only changing implementation details and everything should be covered by tests already.

### Checklist

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

File: pkg/network/buffer.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package network
+
+import "math"
+
+const minBufferSize = 256
+
+// Buffer encapsulates a resizing buffer for providing ConnectionStat pointers
+// to read from the windows driver
```
Comment (brycekahle) [inline pkg/network/buffer.go:8]: nit: no longer windows only 😄

File: pkg/network/buffer.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+type Buffer struct {
```
Comment (brycekahle) [inline pkg/network/buffer.go:9]: nit: `Buffer` is a bit of a generic name. Maybe something like `ConnectionBuffer`?

File: pkg/network/tracer/connection/kprobe/tracer.go [hunk 3]
```diff
@@ -190,29 +181,27 @@ func (t *kprobeTracer) GetMap(name string) *ebpf.Map {
 	}
 }
 
-func (t *kprobeTracer) GetConnections(buffer []network.ConnectionStats, filter func(*network.ConnectionStats) bool) ([]network.ConnectionStats, error) {
+func (t *kprobeTracer) GetConnections(active, closed *network.Buffer, filter func(*network.ConnectionStats) bool) (int, int, error) {
```
Comment (brycekahle) [inline pkg/network/tracer/connection/kprobe/tracer.go:184]: Do we need the separate count return values, couldn't you get the length of the connections slice returned from the buffer? Or even call `Buffer.Len()`?

File: pkg/network/tracer/tracer_test.go [hunk 4]
```diff
@@ -1195,101 +1195,6 @@ func TestConnectedUDPSendIPv6(t *testing.T) {
 	assert.Equal(t, bytesSent, int(outgoing[0].MonotonicSentBytes))
 }
 
-func TestConnectionClobber(t *testing.T) {
```
Comment (brycekahle) [inline pkg/network/tracer/tracer_test.go:1198]: Why did you remove this test? It should be easy to adapt it.

Comment (p-lambert) [reply pkg/network/tracer/tracer_test.go:1198]: sorry, I was actually going to ping you about this to validate whether we still need this or not. Could you elaborate on what we're testing here?

Comment (brycekahle) [reply pkg/network/tracer/tracer_test.go:1198]: This is making sure that a 2nd call to `GetActiveConnections` doesn't destroy/clobber the data returned from the 1st call. The careful choice of the # of connections is to prevent an underlying buffer resize (which definitely wouldn't clobber anything).

Comment (brycekahle) [reply pkg/network/tracer/tracer_test.go:1198]: See https://github.com/DataDog/datadog-agent/pull/6557

File: pkg/network/state.go [hunk 5]
```diff
@@ -208,8 +208,7 @@ func (ns *networkState) GetDelta(
 	}
 	ns.clients[id].stats = newStats
 
-	// Flush closed connection map and stats
-	ns.clients[id].closedConnections = map[string]ConnectionStats{}
+	ns.clients[id].closedConnections = nil
```
Comment (brycekahle) [inline pkg/network/state.go:211]: If we assume `closedConnections` is initialized, isn't this a bad idea to set to `nil`?

File: pkg/network/buffer.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	return &ConnectionBuffer{
+		buf: make([]ConnectionStats, int(math.Max(float64(size), minBufferSize))),
+	}
+}
+
+// Next returns the next `ConnectionStats` object available for writing.
+// It will resize the internal buffer if necessary.
+func (b *ConnectionBuffer) Next() *ConnectionStats {
+	if b.off >= len(b.buf) {
+		b.buf = append(b.buf, ConnectionStats{})
```
[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/network/buffer.go:24]: I'm pretty sure Golang doubles the buffer size by default: https://play.golang.org/p/NapAC5QEwEw
(I also added a benchmark to make sure of that – see `buffer_test.go`)

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/network/buffer.go:24]: Ah yeah, duh!

File: pkg/network/tracer/connection/kprobe/tcp_close_consumer.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+		return
+	}
+	c.perfHandler.Stop()
+}
+
+func (c *tcpCloseConsumer) start() {
+	go func() {
+		for {
+			select {
+			case batchData, ok := <-c.perfHandler.DataChannel:
```
[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/network/tracer/connection/kprobe/tcp_close_consumer.go:96]: Good point. I was also wondering if we could write our own perf poller or add an alternative implementation to the ebpf library that uses a buffer pool, because a large chunk of the memory allocations now come from it.

File: pkg/network/tracer/tracer_test.go (hunk 4 above)
Comment (p-lambert) [reply pkg/network/tracer/tracer_test.go:1198]: I think we could simplify this test and just ensure that the result from `GetConnections` don't point to the same same memory address of `buffer.Connections()`. Would that be ok?

Comment (brycekahle) [reply pkg/network/tracer/tracer_test.go:1198]: That may work, but feels a bit implementation specific to me. 

This test was simply ensuring that the results from two consecutive calls didn't interfere with another. If I could've written it without the sizing logic, I would've, since it makes some assumptions about how the underlying buffer works.

Comment (p-lambert) [reply pkg/network/tracer/tracer_test.go:1198]: yeah, that's why I'm asking. I feel that there is already a sort of implementation leak in the test, and I find it a little bit hard to follow, which is why I'd rather rewrite it in a simpler way.

Comment (p-lambert) [reply pkg/network/tracer/tracer_test.go:1198]: > You could copy the return value of buffer.Connections() into a reused buffer and return that buffer each time. Your test would pass, but it would be clobbering results. 
I think I'm not fully understanding the intent of your test then 😅 
Can you elaborate on what "clobbering results" means in this context?

At the very least it sounds to me that this has only to do with `state.GetDelta` so I could at least remove the cruft of creating a TCP server etc

Comment (brycekahle) [reply pkg/network/tracer/tracer_test.go:1198]: clobbering = the value of the return value from `GetActiveConnections` call #1 changing after `GetActiveConnections` call #2. We are doing lots of tricks to save on allocations, but we need to ensure we aren't modifying values that are in-use elsewhere.

The crux of the test is the assertions:
https://github.com/DataDog/datadog-agent/blob/1e940bafc53b855251b8f19f9ad5d0ed39804bcd/pkg/network/tracer/tracer_test.go#L1288-L1289

The fix was indeed in `state.GetDelta`, but I'm treating the implementation like a blackbox as much as possible.

Comment (p-lambert) [reply pkg/network/tracer/tracer_test.go:1198]:
```
func TestConnectionClobber(t *testing.T) {
	state := newDefaultState()
	active := []ConnectionStats{
		ConnectionStats{
			Source: util.AddressFromString("127.0.0.1"),
			Dest:   util.AddressFromString("8.8.8.8"),
			SPort:  123,
			DPort:  53,
		},
	}

	delta := state.GetDelta("1", 1, active, nil, nil, nil)
	require.Len(t, delta.Connections, 1)

	c := &delta.Connections[0]
	require.Equal(t, 123, int(c.SPort))

	// Modify result value and verify it doesn't interfere to with the original slice
	c.SPort = 456
	assert.Equal(t, 123, int(active[0].SPort))
}
```
Does this look reasonable?

Comment (brycekahle) [reply pkg/network/tracer/tracer_test.go:1198]: So that tests that `active` and `delta` are decoupled, but doesn't test if two successive calls are coupled in some way. (which I think is the largest risk when trying to save allocations. Actually reusing the same memory as what you returned.)

File: pkg/network/tracer/tracer.go [hunk 8]
```diff
@@ -164,25 +159,14 @@ func NewTracer(config *config.Config) (*Tracer, error) {
...
 		ebpfTracer:                 ebpfTracer,
-		closedConnsCh:              closedConnsCh,
 	}
-	tr.initClosedPolling()
 
-	return tr, nil
-}
+	err = ebpfTracer.Start(tr.closedCB)
+	if err != nil {
+		return nil, fmt.Errorf("could not start ebpf manager: %s", err)
```
Comment (brycekahle) [inline pkg/network/tracer/tracer.go:166]: Do we need to tear anything down at this point, since the tracer object has been created along with possible HTTP and DNS features? I think this is why it was before object creation before.

**Review by brycekahle — APPROVED** [review]
One comment, but otherwise looks good.

File: pkg/network/tracer/tracer.go (hunk 8 above)
Comment (p-lambert) [reply pkg/network/tracer/tracer.go:167]: Great catch. I added some changes to handle that scenario
